### PR TITLE
Changes return types of walk engine functions which mutated many references

### PR DESCRIPTION
**Pull request by KipHamiltons**
~`computeCartesianPosition[AtTime]` takes a boolean parameter `isLeftsupportFoot` (passed by value), which is then passed to `TrajectoriesSupportFootState`, which modifies the value as a reference. The call to that function does nothing because the bool is passed by value into the previous function call (this is the only use of `TrajectoriesSupportFootState`).~

~This changes it to be passed by reference, which is certainly the intention~

Update:
I have made the trajectoryutils functions return a tuple instead of mutating its parameters. This is much nicer

### CODE REVIEW

**Comment by TrentHouliston** [comment]
This isn't a very good way to pass this sort of information around in the first place.
Is there a way this can be done without requiring something to be passed by reference?
(via a return or otherwise?)

**Comment by KipHamiltons** [comment]
`TrajectoriesSupportFootState` is designed to set its two boolean references, so we would have to split it into two separate tracks, but I don't see why we couldn't
(assuming we don't want to use `pair` or something, which seems like a terrible idea lol)

**Comment by Bidski** [comment]
> `TrajectoriesSupportFootState` is designed to set its two boolean references, so we would have to split it into two separate tracks, but I don't see why we couldn't
> (assuming we don't want to use `pair` or something, which seems like a terrible idea lol)

I would honestly make it return a pair, I dont see what is horrible about that

**Comment by TrentHouliston** [comment]
Especially with the awesome power of structured bindings in c++17!

**Comment by KipHamiltons** [comment]
oh dang, I didn't realised structured binding was in C++ :heart_eyes: that's one of my favourite things in python

**Review by ysims — COMMENTED** [review]
Make sure this is tested before merging.

**Comment by KipHamiltons** [comment]
I removed `DefaultCheckState` as well, because it was never used

File: module/motion/QuinticWalk/src/WalkEngine.cpp [hunk 1]
```diff
@@ -679,27 +679,21 @@ namespace module::motion {
...
     }
 
-    void QuinticWalkEngine::computeCartesianPosition(Eigen::Vector3f& trunkPos,
-                                                     Eigen::Vector3f& trunkAxis,
-                                                     Eigen::Vector3f& footPos,
-                                                     Eigen::Vector3f& footAxis,
-                                                     bool isLeftsupportFoot) {
-        // Compute trajectories time
-        float time = getTrajsTime();
+    using PositionSupportTuple = std::tuple<Eigen::Vector3f, Eigen::Vector3f, Eigen::Vector3f, Eigen::Vector3f, bool>;
```
Comment (CMurtagh-LGTM) [inline module/motion/QuinticWalk/src/WalkEngine.cpp:682]:
```suggestion
```

File: module/motion/QuinticWalk/src/WalkEngine.cpp [hunk 2]
```diff
@@ -679,27 +679,21 @@ namespace module::motion {
...
 
-        computeCartesianPositionAtTime(trunkPos, trunkAxis, footPos, footAxis, isLeftsupportFoot, time);
+    PositionSupportTuple QuinticWalkEngine::computeCartesianPosition() const {
```
Comment (CMurtagh-LGTM) [inline module/motion/QuinticWalk/src/WalkEngine.cpp:684]:
```suggestion
    QuinticWalkEngine::PositionSupportTuple QuinticWalkEngine::computeCartesianPosition() const {
```

File: module/motion/QuinticWalk/src/WalkEngine.cpp [hunk 3]
```diff
@@ -679,27 +679,21 @@ namespace module::motion {
...
     }
 
-
-    void QuinticWalkEngine::computeCartesianPositionAtTime(Eigen::Vector3f& trunkPos,
-                                                           Eigen::Vector3f& trunkAxis,
-                                                           Eigen::Vector3f& footPos,
-                                                           Eigen::Vector3f& footAxis,
-                                                           bool isLeftsupportFoot,
-                                                           float time) {
+    PositionSupportTuple QuinticWalkEngine::computeCartesianPositionAtTime(const float time) const {
```
Comment (CMurtagh-LGTM) [inline module/motion/QuinticWalk/src/WalkEngine.cpp:690]:
```suggestion
    QuinticWalkEngine::PositionSupportTuple QuinticWalkEngine::computeCartesianPositionAtTime(const float time) const {
```

**Review by CMurtagh-LGTM — CHANGES_REQUESTED** [review]

File: module/motion/QuinticWalk/src/WalkEngine.cpp (hunk 2 above)
Comment (TrentHouliston) [reply module/motion/QuinticWalk/src/WalkEngine.cpp:684]: Are these needed to make it compile?

Comment (CMurtagh-LGTM) [reply module/motion/QuinticWalk/src/WalkEngine.cpp:684]: If we remove the duplicate using statement it shouldn't compile.

Comment (CMurtagh-LGTM) [reply module/motion/QuinticWalk/src/WalkEngine.cpp:684]: This suggestion is in relation to the previous one.

**Comment by ysims** [comment]
> Make sure this is tested before merging.

I've just checked now and it looks fine, doesn't look like there is any difference...

File: shared/utility/motion/splines/TrajectoryUtils.hpp [hunk 4]
```diff
@@ -189,111 +194,68 @@ namespace utility::motion::splines {
...
-        }
-        if (trunkPos.z() - footPos.z() < 0.20f) {
-            cost += 1000.0f - 1000.0f * (trunkPos.z() - footPos.z() - 0.20f);
-        }
-        if (footAxis.norm() >= M_PI_2) {
-            cost += 1000.0f + 1000.0f * (footAxis.norm() - M_PI_2);
-        }
-
-        return cost;
+    [[nodiscard]] inline std::pair<bool, bool> TrajectoriesSupportFootState(float t, const Trajectories& traj) {
```
Comment (TrentHouliston) [inline shared/utility/motion/splines/TrajectoryUtils.hpp:255]: Why is this function PascalCase?

File: shared/utility/motion/splines/TrajectoryUtils.hpp [hunk 5]
```diff
@@ -189,111 +194,68 @@ namespace utility::motion::splines {
      * Compute from given spline container trajectory Cartesian trunk and foot position/velocity/acceleration
      * and assign it to given vector
      */
-    inline void TrajectoriesTrunkFootPos(float t,
-                                         const Trajectories& traj,
-                                         Eigen::Vector3f& trunkPos,
-                                         Eigen::Vector3f& trunkAxis,
-                                         Eigen::Vector3f& footPos,
-                                         Eigen::Vector3f& footAxis) {
+    [[nodiscard]] inline Vector3fQuadruple TrajectoriesTrunkFootPos(const float t, const Trajectories& traj) {
```
Comment (CMurtagh-LGTM) [inline shared/utility/motion/splines/TrajectoryUtils.hpp:197]:
```suggestion
    [[nodiscard]] inline Vector3fQuadruple trajectoriesTrunkFootPos(const float t, const Trajectories& traj) {
```

File: shared/utility/motion/splines/TrajectoryUtils.hpp [hunk 6]
```diff
@@ -189,111 +194,68 @@ namespace utility::motion::splines {
...
+        return {trunkPosVel, trunkAxisVel, footPosVel, footAxisVel};
     }
 
-    inline void TrajectoriesTrunkFootAcc(float t,
-                                         const Trajectories& traj,
-                                         Eigen::Vector3f& trunkPosAcc,
-                                         Eigen::Vector3f& trunkAxisAcc,
-                                         Eigen::Vector3f& footPosAcc,
-                                         Eigen::Vector3f& footAxisAcc) {
+    [[nodiscard]] inline Vector3fQuadruple TrajectoriesTrunkFootAcc(const float t, const Trajectories& traj) {
```
Comment (CMurtagh-LGTM) [inline shared/utility/motion/splines/TrajectoryUtils.hpp:231]:
```suggestion
    [[nodiscard]] inline Vector3fQuadruple trajectoriesTrunkFootAcc(const float t, const Trajectories& traj) {
```

File: shared/utility/motion/splines/TrajectoryUtils.hpp [hunk 7]
```diff
@@ -189,111 +194,68 @@ namespace utility::motion::splines {
...
+        return {trunkPos, trunkAxis, footPos, footAxis};
     }
 
-    inline void TrajectoriesTrunkFootVel(float t,
-                                         const Trajectories& traj,
-                                         Eigen::Vector3f& trunkPosVel,
-                                         Eigen::Vector3f& trunkAxisVel,
-                                         Eigen::Vector3f& footPosVel,
-                                         Eigen::Vector3f& footAxisVel) {
+    [[nodiscard]] inline Vector3fQuadruple TrajectoriesTrunkFootVel(const float t, const Trajectories& traj) {
```
Comment (CMurtagh-LGTM) [inline shared/utility/motion/splines/TrajectoryUtils.hpp:214]:
```suggestion
    [[nodiscard]] inline Vector3fQuadruple trajectoriesTrunkFootVel(const float t, const Trajectories& traj) {
```

File: shared/utility/motion/splines/TrajectoryUtils.hpp (hunk 4 above)
Comment (CMurtagh-LGTM) [reply shared/utility/motion/splines/TrajectoryUtils.hpp:255]:
```suggestion
    [[nodiscard]] inline std::pair<bool, bool> trajectoriesSupportFootState(float t, const Trajectories& traj) {
```

**Review by CMurtagh-LGTM — CHANGES_REQUESTED** [review]

**Review by CMurtagh-LGTM — APPROVED** [review]
LGTM

**Pull request merged**